client.go

```diff
@@ -0,0 +1,179 @@
+	Log   types.FluentBitLog
+	Error string
+}
+
+type RawLogResult struct {
+	Result any    `json:"result,omitempty"`
+	Error  string `json:"error,omitempty"`
+}
+
+type response struct {
+	JsonRpcVersion string         `json:"jsonrpc"`
+	Id             int            `json:"id"`
+	Result         []RawLogResult `json:"result,omitempty"`
+	Error          *rpcError      `json:"error,omitempty"`
+}
+
+type rpcError struct {
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+}
+
+func New(url string) *Client {
+	rv := &Client{
+		url:    url,
+		nextId: 1,
+		httpClient: http.Client{
+			Transport: &http.Transport{
+				MaxIdleConns:       10,
+				IdleConnTimeout:    30 * time.Second,
+				DisableCompression: true,
+			},
+		},
+	}
+
+	return rv
+}
+
+func (c *Client) Run(events []types.FluentBitLog, filter string) ([]LogResult, error) {
+	id := c.nextId
+	c.nextId += 1
+
+	eventAttrs := []types.FluentBitLogAttrs{}
+	for _, e := range events {
+		eventAttrs = append(eventAttrs, e.Attrs)
+	}
+
+	reqBody, err := json.Marshal(&request{
+		JsonRpcVersion: "2.0",
+		Id:             id,
+		Method:         "run",
+		Params: params{
+			Events: eventAttrs,
+			Filter: filter,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	httpRes, err := c.httpClient.Post(c.url, "application/json", bytes.NewReader(reqBody))
+	if err != nil {
+		return nil, err
+	}
+	defer httpRes.Body.Close()
+
+	resBody, err := io.ReadAll(httpRes.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if httpRes.StatusCode >= 400 {
+		return nil, fmt.Errorf("HTTP Error (%v): %v", httpRes.Status, string(resBody))
+	}
+
+	var response response
+	err = json.Unmarshal(resBody, &response)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to unmarshal json response: %v", string(resBody))
+	}
+
+	if response.Id != id {
+		return nil, fmt.Errorf("Unexpected response id: %v", string(resBody))
+	}
+
+	if response.JsonRpcVersion != "2.0" {
+		return nil, fmt.Errorf("Unexpected response RPC version: %v", string(resBody))
+	}
+
+	if response.Error != nil {
+		return nil, fmt.Errorf("RPC call error: %v (%v)", response.Error.Code, response.Error.Message)
+	}
+
+	rv := []LogResult{}
+	for i, r := range response.Result {
+		if r.Error != "" {
+			rv = append(rv, LogResult{Error: r.Error})
+			continue
+		}
+
+		resultItems := r.Result.([]any)
+		if len(resultItems) != 3 {
+			return nil, fmt.Errorf("RPC call returned unexpected result (wrong number of items)")
+		}
+
+		code := resultItems[0].(float64)
+		if code == -1 {
+			// drop record
+			continue
+		}
+
+		if code == 0 {
+			// record not modified, use the original
+			rv = append(rv, LogResult{Log: events[i]})
+			continue
+		}
+
+		var timestamp types.FluentBitTime
+		if code == 2 {
+			// use the input timestamp
+			timestamp = events[i].Timestamp
+		} else {
+			timestamp = types.FluentBitTime(resultItems[1].(float64))
+		}
+
+		attrs := resultItems[2]
+		if reflect.TypeOf(attrs).Kind() == reflect.Slice {
```

Comment:
I think this is doing too much. It's just a client for JSONRPC.
It should not have this logic were it checks the return code, skip, modify, or add records...

Reply:
Here's a few reasons I haven't added this logic to the server:

- It is designed to be a simple way to run a Lua function on a list of objects and return the results, it is more flexible and reusable this way. (That's why I was against adding processing rules knowledge to the server).

- This logic is all part of fluent-bit lua interface and Eduardo already mentioned in a call this interface it can change, if we run Lua code on input/output plugins for example. If we have a different interface for input plugins, it will make the server even more complex.

- The last reason to maintenance: It is much more difficult to maintain/test this on a C codebase than it is on a golang codebase. 

With that said, I agree that this is not ideal and I'm not fully against moving this logic to the server once the Lua interface is stable and we have determined if processing rules will indeed be running on filters or on input plugins. I just don't think there's any benefit to doing this now, and it will only delay delivering processing rules.

Reply:
It's find then. Let's keep the Lua server just to execute Lua.
But I think this part could be done differently without reflect. I might send you a PR later.